### ISSUE TITLE
Better support for monorepos

### ISSUE THREAD

**Issue by paeolo** [ds300/patch-package#277]
# 🐛 bug report

On a lerna-based monorepo, there are some issues with patch-package.

Suppose I have this folder structure...

```
.
├── ...
├── packages
│   ├── a
│     └── patches
│     └── package.json
├── node_modules
└── package.json
```
... and suppose I put patch-package dependency in the package `a` together with a `postinstall` script.
## 🤔 Current Behavior
The package `a` should work in isolation from the top-level folder and other packages.

- When I do a `npm i` on the top-level folder, it will install dependencies and so do a npm i on `a` (if things are well configured). But probably it will `hoist` dependencies of `a` in the top-level `node_modules` instead of a local one (ie. instead of a `node_modules` inside the `a` folder).
- So `patch-package` will be executed and tries to patch things in `a/node_modules` but since dependencies are hoisted in the top-level folder, it will fail (if no symlink are present).

## 💁 Possible Solution

Add a flag to enable patching of the top-level `node_modules`. ie: 
- it's very easy to find the root folder, just do a `find-up` for a `package.json` that contains a `workspaces` field and/or a `lerna.json` file.
- From there, if you can't patch the local `node_modules` folder (ie. for instance it doesn't exist) tries to patch the root `node_modules` (when flag is enabled)

## 🔦 Context

- I have a monorepo and I really don't want to put the patches in the root folder, but rather in the packages, so patch-package can work for packages individually.
- Also in my setup there are no local `node_modules` ie. no `node_modules` in `packages/a` and I would like patch-package to work out-of-the-box for this kind of situation (with a flag if it's ok).
- I don't want to patch `patch-package`

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| patch-package           | 6.2.2

**Comment by jony89** [comment]
+1 

Experiencing the exact same issue. this becomes even more needed since npm 7 workspaces is out

**Comment by KevinGorjan** [comment]
+1

**Comment by tappin-kr** [comment]
Any update on this? Currently running into this same issue, its a pain.

Any workarounds? @paeolo ?

**Comment by tappin-kr** [comment]
As a hacky workaround, I was able to add this to my postinstall.
```js
"postinstall": "cp -r ../../adal-node node_modules/adal-node && patch-package",
```

**Comment by paeolo** [comment]
@tappin-kr I didn't use it yet, but I planned to use Yarn 2 and it seems it can handle patches itself without the need for an external package like patch-package

**Comment by wibb36** [comment]
@paeolo Yes, sure!! but how about those who are not planning on migrating to Yarn 2 just yet(such as myself)? Monorepo is the way to go these days and it would be nice to have patch-package supports it.

**Comment by JB-CHAUVIN** [comment]
Hello, anything new about this issue ?

**Comment by bd82** [comment]
I am wondering if this is an issue with `patch-package` or if this could be solved if and when npm would
support the `nohoist` option?
- https://github.com/npm/rfcs/issues/287

**Comment by 0x0a0d** [comment]
my solution: https://gist.github.com/0x0a0d/a22bdb57c736c8f0adc9209295aff7ed

**Comment by NathanC** [comment]
Any updates or modern workarounds on this? The solution that @0x0a0d gave isn't relevant to me as I'm using the new version of lerna that simply uses workspaces.

**Comment by chuyik** [comment]
@NathanC use pnpm or yarn berry if you could, which provides the `patch` support.
https://pnpm.io/cli/patch

**Comment by tcguoxing** [comment]
I also have the same problem and my solution is to set the "--path-dir" as the relative path value after my "patch-package" command.

**Comment by asp3** [comment]
having the same issue, using turbo repo.

**Comment by carlmagumpara** [comment]
+

**Comment by agusting19** [comment]
+1

**Comment by leochen-g** [comment]
+1

**Comment by sergeyshmakov** [comment]
Workaround this issue:
* Place command `"patch:packageA": "patch-package"` in root `package.json`
* Place hook command in `packageA` folder `"postinstall": "cd ../ && npm run patch:packageA"`

**Comment by Monte9** [comment]
@sergeyshmakov your solution worked! Thanks a lot, you saved me a lot of time 🎉

**Comment by Dimon70007** [comment]
OK. what behaviour do you expect for situation when two packages contain patch for the same library? What patch should been applied in this case?

**Comment by LEI** [comment]
Maybe both patches could be applied especially if they do not affect the same part of the file?

Otherwise I would expect the monorepo to have priority over project specific patches, with a warning or an error in case of conflict.

**Comment by BryceLTaylor** [comment]
Related to this issue, I filed this issue with NPM for how postinstall works with their worspaces feature: https://github.com/npm/cli/issues/8442

**Comment by manzoorwanijk** [comment]
Because of this issue (#277) and #595, I [proposed](https://github.com/npm/rfcs/pull/862) and [implemented](https://github.com/npm/cli/pull/9439) the `npm patch` feature in npm CLI itself, and it will be a part of [npm v12](https://github.com/npm/cli/pull/9387). So, we no longer need this package after npm v12.
